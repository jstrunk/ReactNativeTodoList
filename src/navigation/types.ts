import { StackNavigationProp } from "@react-navigation/stack";
import { RouteProp } from "@react-navigation/native";

export type RootStackParamList = {
  TodoSelect: undefined;
  TodoList: {id: string};
};

export type TodoSelectScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'TodoSelect'
>;

export type TodoListScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'TodoList'
>;

export type TodoListScreenRouteProp = RouteProp<
  RootStackParamList,
  'TodoList'
>;

