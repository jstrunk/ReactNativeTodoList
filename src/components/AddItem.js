import React, {Component} from 'react';
import {TextInput,Text,StyleSheet} from 'react-native';

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  addItem: {
    padding: 15,
    textAlign: 'left',
    color: 'grey',
    borderBottomColor: 'lightgrey',
    borderBottomWidth: 1,
  },
});

export default class AddItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      newItem: '',
    }
  }

  onChangeText = (text) => {
    this.setState({newItem: text});
  }

  addItem = () => {
    if (this.state.newItem == '') return;

    let items = [...this.props.items];
    items.push({key: this.state.newItem, done: false});
    this.props.onUpdate(items);
    this.setState({newItem: ''});
    this.addItemInput.setNativeProps({text: ''});
  }

  render() {

    return (
      <TextInput
        ref={addItem => this.addItemInput = addItem }
        style={styles.addItem}
        onChangeText={this.onChangeText}
        onSubmitEditing={this.addItem}
        placeholder="Add new item"
      />
    );
  }
}