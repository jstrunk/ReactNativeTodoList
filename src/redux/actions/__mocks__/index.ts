export const addItem = jest.fn();
export const toggleItem = jest.fn();
export const deleteItem = jest.fn(); 
export const deleteCompleted = jest.fn(); 
export const reorderList = jest.fn(); 
export const registerList = jest.fn(); 
export const setActiveList = jest.fn(); 
export const createList = jest.fn(); 
export const loadList = jest.fn(); 