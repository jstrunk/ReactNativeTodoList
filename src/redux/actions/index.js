import { ADD_ITEM, DELETE_ITEM, TOGGLE_ITEM, DELETE_COMPLETED } from '../constants';

export function addItem(text) {
  return { type: ADD_ITEM, text };
}

export function toggleItem(index) {
  return { type: TOGGLE_ITEM, index };
}

export function deleteItem(index) {
  return { type: DELETE_ITEM, index };
}

export function deleteCompleted() {
  return { type: DELETE_COMPLETED };
}
