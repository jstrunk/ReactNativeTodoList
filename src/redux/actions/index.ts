import { ADD_ITEM,
  DELETE_ITEM,
  TOGGLE_ITEM,
  DELETE_COMPLETED,
  TodoActionType, } from '../types';

export function addItem(text: string): TodoActionType {
  return { type: ADD_ITEM, text };
}

export function toggleItem(index: number): TodoActionType {
  return { type: TOGGLE_ITEM, index };
}

export function deleteItem(index: number): TodoActionType {
  return { type: DELETE_ITEM, index };
}

export function deleteCompleted(): TodoActionType {
  return { type: DELETE_COMPLETED };
}
