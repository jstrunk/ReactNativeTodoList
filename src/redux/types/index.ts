export interface Todo {
  key: string;
  done: boolean;
}

export interface ITodoList {
  byId: Record<string,Todo>;
  allItems: Array<string>;
}

export interface ITodoState {
  todoList: ITodoList;
}

export const defaultState: ITodoList = {
  byId: {},
  allItems: [],
};

export const ADD_ITEM = "ADD_ITEM";
export const TOGGLE_ITEM = "TOGGLE_ITEM";
export const DELETE_ITEM = "DELETE_ITEM";
export const DELETE_COMPLETED = "DELETE_COMPLETED";

interface AddItemAction {
  type: typeof ADD_ITEM;
  text: string;
  id: string;
}

interface ToggleItemAction {
  type: typeof TOGGLE_ITEM;
  id: string;
}

interface DeleteItemAction {
  type: typeof DELETE_ITEM;
  id: string;
}

interface DeleteCompletedAction {
  type: typeof DELETE_COMPLETED;
  ids: Array<string>;
}

export type TodoActionType = AddItemAction | ToggleItemAction |
  DeleteItemAction | DeleteCompletedAction;
