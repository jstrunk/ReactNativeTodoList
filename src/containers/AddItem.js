import React, {Component} from 'react';
import {TextInput,Text,StyleSheet} from 'react-native';
import { connect } from 'react-redux'
import { addItem } from '../redux/actions'

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  addItem: {
    padding: 15,
    textAlign: 'left',
    color: 'grey',
    borderBottomColor: 'lightgrey',
    borderBottomWidth: 1,
  },
});

class AddItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      newItem: '',
    }
  }

  onChangeText = (text) => {
    this.setState({newItem: text});
  }

  submitItem = () => {
    if (this.state.newItem == '') return;
    this.props.addItem(this.state.newItem);
    this.setState({newItem: ''});
    this.addItemInput.setNativeProps({text: ''});
  }

  render() {

    return (
      <TextInput
        ref={submitItem => this.addItemInput = submitItem }
        style={styles.addItem}
        onChangeText={this.onChangeText}
        onSubmitEditing={this.submitItem}
        placeholder="Add new item"
      />
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addItem: text => dispatch(addItem(text)),
})

export default connect(
  null,
  mapDispatchToProps
)(AddItem)