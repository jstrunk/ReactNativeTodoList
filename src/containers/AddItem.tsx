import React, {Component} from 'react';
import {TextInput,StyleSheet} from 'react-native';
import { connect, ConnectedProps } from 'react-redux'
import { addItem } from '../redux/actions'
import { ITodoState } from '../redux/types';
import itemId from '../lib/itemId';

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  addItem: {
    padding: 15,
    textAlign: 'left',
    color: 'grey',
    borderBottomColor: 'lightgrey',
    borderBottomWidth: 1,
  },
});

const mapStateToProps = (state: ITodoState) => ({
  order: state.todoList.allItems,
})

const mapDispatch = {
  addItem: (text: string, id: string) => addItem(text, id),
}

const connector = connect(
  mapStateToProps,
  mapDispatch
);

type PropsFromRedux = ConnectedProps<typeof connector>;

type AddItemState = {
  newItem: string;
};

class AddItem extends Component<PropsFromRedux, AddItemState> {
  addItemInput: React.RefObject<TextInput>;
  constructor(props: PropsFromRedux) {
    super(props);
    this.state = {
      newItem: '',
    };
    this.addItemInput = React.createRef<TextInput>();
  }

  onChangeText = (text: string) => {
    this.setState({newItem: text});
  }

  submitItem = async () => {
    if (this.state.newItem == '') return;
    let newId = await itemId(this.state.newItem);
    if (this.props.order.includes(newId)) {
      newId = await itemId(this.state.newItem + '1');
    }
    this.props.addItem(this.state.newItem, newId);
    this.setState({newItem: ''});
    if (this.addItemInput !== null && this.addItemInput.current !== null) {
      this.addItemInput.current.setNativeProps({text: ''});
    }
  }

  render() {
    return (
      <TextInput
        ref={this.addItemInput}
        style={styles.addItem}
        onChangeText={this.onChangeText}
        onSubmitEditing={this.submitItem}
        placeholder="Add new item"
      />
    );
  }
}

export default connector(AddItem)
