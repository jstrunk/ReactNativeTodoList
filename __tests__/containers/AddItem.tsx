import 'react-native';
import React from 'react';
import AddItem from '../../src/containers/AddItem';
import * as types from '../../src/redux/types';
import * as actions from '../../src/redux/actions';
import { Provider } from 'react-redux';
import { render, fireEvent } from '@testing-library/react-native'
import configureMockStore, { MockStore } from 'redux-mock-store';
import thunk, { ThunkDispatch } from 'redux-thunk';
import { oneListOneItemDone } from '../../test/testStates';

jest.mock('../../src/lib/itemId');

jest.mock('../../src/redux/actions');

type DispatchExts = ThunkDispatch<types.ITodoState, void, types.CombinedActionType>;
const middlewares = [thunk];
const mockStore = configureMockStore<types.ITodoState,DispatchExts>(middlewares);

describe('AddItem container', () => {
  let store: MockStore;

  beforeEach(() => {
    store = mockStore(oneListOneItemDone);

    store.dispatch = jest.fn();
  });

  it('renders an AddItem', () => {
    const { toJSON } = render(
      <Provider store={store}>
        <AddItem />
      </Provider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should dispatch an add item action', async () => {
    const { getByPlaceholderText } = render(
      <Provider store={store}>
        <AddItem />
      </Provider>
    );

    const element = getByPlaceholderText('Add new item');

    fireEvent.changeText(element, 'Hello World!');
    fireEvent(element, 'submitEditing');

    await new Promise((r) => setTimeout(r, 2000));

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      actions.addItem('Hello World!', 'Hello World!')
    );
  });

  it('should handle id collisions', async () => {
    const { getByPlaceholderText } = render(
      <Provider store={store}>
        <AddItem />
      </Provider>
    );

    const element = getByPlaceholderText('Add new item');

    fireEvent.changeText(element, 'def2');
    fireEvent(element, 'submitEditing');

    await new Promise((r) => setTimeout(r, 2000));

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      actions.addItem('def2', 'def21')
    );
  });

  it('should not dispatch an add item action on empty input', () => {
    const { getByPlaceholderText } = render(
      <Provider store={store}>
        <AddItem />
      </Provider>
    );

    const element = getByPlaceholderText('Add new item');

    fireEvent.changeText(element, '');
    fireEvent(element, 'submitEditing');

    expect(store.dispatch).toHaveBeenCalledTimes(0);
  });
});
