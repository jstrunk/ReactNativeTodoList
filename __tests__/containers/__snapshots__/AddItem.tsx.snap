// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddItem container renders an AddItem 1`] = `
<TextInput
  allowFontScaling={true}
  onChangeText={[Function]}
  onSubmitEditing={[Function]}
  placeholder="Add new item"
  rejectResponderTermination={true}
  style={
    Object {
      "borderBottomColor": "lightgrey",
      "borderBottomWidth": 1,
      "color": "grey",
      "padding": 15,
      "textAlign": "left",
      "width": "100%",
    }
  }
  underlineColorAndroid="transparent"
/>
`;
